klog.go

```diff
@@ -812,14 +812,17 @@ func (l *loggingT) output(s severity, log logr.InfoLogger, buf *buffer, file str
 		}
 	}
 	data := buf.Bytes()
-	if log != nil {
-		// TODO: set 'severity' and caller information as structured log info
-		// keysAndValues := []interface{}{"severity", severityName[s], "file", file, "line", line}
+	// In later version, we should check argument s is either errorLog or infoLog.
+	// However for now, as backward compatibility, all s other than errorLog goes to l.logr.Info().
+	// if l.logr != nil && (s == errorLog || s == infoLog) {
+	if l.logr != nil {
 		if s == errorLog {
-			l.logr.Error(nil, string(data))
+			l.logr.Error(nil, string(data), "severity", severityName[s], "file", file, "line", line)
 		} else {
-			log.Info(string(data))
+			l.logr.Info(string(data), "severity", severityName[s], "file", file, "line", line)
 		}
+	} else if log != nil && s == infoLog {
```

Comment:
It should technically never be possible to get into a state where `log` == nil, l.logr != nil and `s == errorLog` right? Because there is no `Error` method on an InfoLogger. I ask because otherwise there is a risk we drop errors here 😄 

This logic looks right to me :) just trying to validate my own assumptions!

Reply:
Thanks @munnerz .

Yes, that is underlying idea of this line. Ideally, it should not happen.

Reply:
Makes sense. If `l.logr == nil`, `log != nil` and `s` is *not* `infoLog`, it seems like log messages will be silently dropped. I'm not sure if this is a feasible situation to be in, but with the previous code we'd *always* log unconditionally if `log != nil`. Is there a reason for the `s == infoLog` check on this line?

Reply:
Yes, as you said, if `l.logr == nil`, `log != nil` and `s` is not `infoLog`, log messages will be dropped.
I read [`logr` package's `README`](https://github.com/go-logr/logr) and thought it's good approach that `klog` only handles `infoLog` or `errorLog` when `logr.InfoLogger` or `logr.Logger` is used. Because user of `logr.InfoLogger` & `logr.Logger` only expects **Info** or **Error**, I think.
So, future development of `klog` should care about this `logr`'s idea.

As I wrote in `klog_test.go`, currently this commit doesn't hurt `klog` package's code. And there is no public method/function with which user can specify `severity` value of `output` method.

Reply:
`Because user of logr.InfoLogger & logr.Logger only expects Info or Error, I think.`

There are log levels in between (e.g. warning) that do not have direct equivalents in a logr world (as logr only has levels). The thinking is to treat any un-levelled log message that specifies a non-info level severity as an Info message and pass along the `severity` as a 'key/value' in the log line (i.e. `severity=warn`) (except for Errors, which are always passed to the root Logger via the Error method). The reason this severity k/v pair wasn't added when `SetLogger` was originally added can be seen here: https://github.com/kubernetes/klog/pull/20#issuecomment-469389056

Mostly there wasn't clear consensus as to what 'keys' should be used and what the onus is on implementors of the `logr` interface wrt setting these keys so other consumers can consume them reliably 😄

Reply:
I understand why k/v isn't added. It's hard problem, until some set of standard keys is established.

Thanks!